### ISSUE TITLE
support for x11mod kernel

### ISSUE THREAD

**Issue by t0mpr1c3** [nicehash/sph-sgminer#1]
I would like to see support for x11mod kernel, I am working on it currently and will submit a pull request if I get somewhere with it.

**Comment by kenshirothefist** [comment]
This is great. However, please, contact MrBrdo https://github.com/mrbrdo, who is also working on this issue and will port this to official sgminer - maybe you can split work or work together!

**Comment by mrbrdo** [comment]
I already have full X11-mod support, just needs more testing and a weird bug to fix. If you wanna help let me know.

**Comment by t0mpr1c3** [comment]
Cool. Happy to test.

**Comment by t0mpr1c3** [comment]
My own version - apparently working, although there were some compile warnings. https://github.com/t0mpr1c3/sph-sgminer

**Comment by bitbandi** [comment]
Hi, Just some note: 
- pls correct code indent at newly added lines. (lasybear sometime use space sometimes use tabs, grr...)
- There are some unnessesary files (like m4/*), remove it! (builder will copy the own version if need)
- There are some dead code, remove it!
  
  ```
  #define CL_CHECK ...
  #define CL_CHECK_ERR ...
  ```
- compare the code with nicehash codebase, less codechange less bugs!
  https://github.com/t0mpr1c3/sph-sgminer/compare/nicehash:master...master

**Comment by mkminer** [comment]
t0mpr1c3,

Looks like you have a bad merge of the "configure" file. I just ran:
 CFLAGS="-O2 -Wall -march=native" ./configure

And got this:

./configure: line 16102: syntax error near unexpected token `<<<'
./configure: line 16102:`<<<<<<< HEAD'

Mark Spehalski
markspehalski@gmail.com
503-484-3539

**Comment by t0mpr1c3** [comment]
Mark, I fixed up the configure.ac file, there is a more recent commit that I had not pushed.

**Comment by mrbrdo** [comment]
I suggest you wait a couple of days until I release mine, it supports all kernels from sph-sgminer, X11-mod and also Keccak (Maxcoin), on top of that it supports hot-swapping the kernels and it is completely based on latest sgminer. Also @bitbandi thanks for the base algorithm structure that you made, it helped to kickstart it a lot.
It will be in the official sgminer-dev/sgminer repo.

**Comment by t0mpr1c3** [comment]
That is cool, this version is really just for me over the next couple of days so that I can use both x11mod kernel and retain the proper pool idle behaviour for Nicehash which I believe is not implemented in the lasybear code.

**Comment by mrbrdo** [comment]
Sure, no harm in that :)
By the way if any of you guys have any idea, the only issue I am having is when ("hot") switching from darkcoin to scrypt, but only when starting out with darkcoin, I will get invalid nonces on scrypt (over diff1targ - 0xffff) until a new block is set. I am pretty sure it is something from the existing codebase (either sph-sgminer or sgminer), but it only happens when doing a hot switch of the kernel/pool. It is 100% reproducible, and it does not happen when switching from scrypt to darkcoin (and I can then switch back to scrypt and this problem is not there). The kernel itself is not touched between the invalid nonces and after it starts working, so the bug is very probably not in CL code, but it could be in the parameters to CL code or buffer state, somehow.

Any idea what could be causing it?

**Comment by mrbrdo** [comment]
Found it. :) So random. And yeah it was in existing code from even cgminer 3.7.2... Just was not meant for hot switching kernels with such different difficulties. Got it working now, will be done soon.

**Comment by bitbandi** [comment]
@mrbrdo , i see keccak256 hash function in sphlib. Can we use that function? and not merging a external implementation?

**Comment by mrbrdo** [comment]
Nope. The maxcoin PoW has quite a few other changes (e.g. FOUND is 0x0F and not 0xFF, etc). But it's not too bad.

**Comment by mrbrdo** [comment]
Btw @bitbandi , is it possible to get a hold of you on IRC or Skype? I will be merging most of your nicehash related fixes into sgminer, I already did some a few weeks ago but for some I was not exactly sure what the change was for, I don't want to merge something I do not understand, so it would be nice I could throw a few questions to you. Also am a bit confused which are relevant since you have multiple repositories :)

**Comment by Artic2017** [comment]
hbkh
